components/core/tests/test-NetworkReader.cpp

```diff
@@ -203,26 +206,44 @@ TEST_CASE("network_reader_with_valid_http_header_kv_pairs", "[NetworkReader]") {
                 fmt::format("Unit-Test-Value{}", i)
         );
     }
-    clp::NetworkReader reader{
-            "https://httpbin.org/headers",
-            0,
-            false,
-            clp::CurlDownloadHandler::cDefaultOverallTimeout,
-            clp::CurlDownloadHandler::cDefaultConnectionTimeout,
-            clp::NetworkReader::cDefaultBufferPoolSize,
-            clp::NetworkReader::cDefaultBufferSize,
-            valid_http_header_kv_pairs
-    };
-    auto const content{get_content(reader)};
-    REQUIRE(assert_curl_error_code(CURLE_OK, reader));
-    auto const parsed_content = nlohmann::json::parse(content);
-    auto const& headers{parsed_content.at("headers")};
-    for (auto const& [key, value] : valid_http_header_kv_pairs) {
-        REQUIRE((value == headers.at(key).get<std::string_view>()));
+    bool is_pass{false};
+    for (size_t i{0}; i < cNumMaxTrials; ++i) {
+        clp::NetworkReader reader{
+                "https://httpbin.org/headers",
+                0,
+                false,
+                clp::CurlDownloadHandler::cDefaultOverallTimeout,
+                clp::CurlDownloadHandler::cDefaultConnectionTimeout,
+                clp::NetworkReader::cDefaultBufferPoolSize,
+                clp::NetworkReader::cDefaultBufferSize,
+                valid_http_header_kv_pairs
+        };
+        auto const content{get_content(reader)};
+        if (false == assert_curl_error_code(CURLE_OK, reader)) {
+            continue;
+        }
+        auto const parsed_content = nlohmann::json::parse(content);
+        auto const& headers{parsed_content.at("headers")};
+        bool are_headers_same{true};
+        for (auto const& [key, value] : valid_http_header_kv_pairs) {
+            if (value != headers.at(key).get<std::string_view>()) {
+                are_headers_same = false;
+                break;
+            }
```

Comment:
We should assert the returned headers are the same if we get a valid response from the server right?

Reply:
I think don't need to asset here:
+ If not the same, it breaks the for loop then `are_headers_same` will be false, then continue to the next trial
+ If using assert here, then there won't be next trial any more, it will exit due to assertion failure exception

Reply:
But exiting due to assertion failure should be the expected behaviour in this case.
Adding this retry is to avoid server-end errors, which is caught in the previous `continue` already.
When the control flow reaches here, the server's response should be successful. If then we have unmatched headers, it's no longer the server's network issue, and we shouldn't swallow this mismatch.